sdk/common/src/main/java/io/opentelemetry/sdk/resources/Resource.java

```diff
@@ -155,8 +161,29 @@ public Resource merge(@Nullable Resource other) {
 
     AttributesBuilder attrBuilder = Attributes.builder();
     attrBuilder.putAll(other.getAttributes());
-    attrBuilder.putAll(this.getAttributes());
-    return new AutoValue_Resource(attrBuilder.build());
+
+    Attributes currentAttributes = this.getAttributes();
+    // a little hack to make sure that the child service.name can override the fallback.
+    String otherServiceName = other.getAttributes().get(ResourceAttributes.SERVICE_NAME);
+    if (!FALLBACK_SERVICE_NAME.equals(otherServiceName)) {
+      currentAttributes =
+          currentAttributes.toBuilder()
+              .put(ResourceAttributes.SERVICE_NAME, otherServiceName)
+              .build();
+    }
+
+    attrBuilder.putAll(currentAttributes);
+    return createResource(attrBuilder.build());
+  }
+
+  private static AutoValue_Resource createResource(Attributes attributes) {
```

Comment:
I don't think we should add the resource attributes everywhere. The service name should be treated just the same as the telemetry.sdk.name, which is also (slated to be) "always" required.

Reply:
If I create a resource with certain attributes explicitly without merging the default resource, I would be very surprised to have any additional attributes. And it seems to make all the code more complicated.

Reply:
The spec requires it now. I think you have a disagreement with the spec, not with the implementation?

Reply:
This is not how I interpret the spec. I think it would be enough if every TracerProvider / Metric merged in a resource with the service name. But even that is quite a strict interpretation. I think it would be enough if it is in the default resource.

@carlosalberto You wrote this spec, WDYT?

Reply:
I created https://github.com/open-telemetry/opentelemetry-specification/pull/1294

Reply:
If you merged the default resource with one that contains the service.name, the new one would not get picked up, right?

Reply:
Correct. Normally you would instead merge your custom resource with the default resource instead of the other way round (the left side is preferred, the right side is supplementary).